### ISSUE TITLE
ASN1Writer does unsafe things with GetLength on TLVReader

### ISSUE THREAD

**Issue by bzbarsky-apple** [project-chip/connectedhomeip#9299]
#### Problem
`ASN1Writer::PutBitString(uint8_t unusedBitCount, chip::TLV::TLVReader & encodedBits)` and `ASN1Writer::PutValue(uint8_t cls, uint32_t tag, bool isConstructed, chip::TLV::TLVReader & val)` have similar code that looks dangerous to me.  Quoting from `PutValue`:
```
    valLen = val.GetLength();

    ReturnErrorOnFailure(EncodeHead(cls, tag, isConstructed, valLen));

    val.GetBytes(mWritePoint, valLen);
    mWritePoint += valLen;

    return CHIP_NO_ERROR;
```

What happens if the TLV is malicious or malformed and `GetLength()` returns a length that is larger than the amount of data actually available in the reader?  We will encode that length, then the `GetBytes` call will fail (possibly after copying some number of bytes smaller than `valLen`), then we will adjust our write point and return success.  But we won't actually have initialized all the bytes we are claiming to be putting into our buffer.

Also, the use of `EncodeHead` here is unsafe and can lead to a buffer overrun with malicious TLV.  The way `EncodeHead` works is that it does:
```
    totalLen = 1 + bytesForLen + (len != kUnknownLength ? len : 0);
    VerifyOrReturnError((mWritePoint + totalLen) <= mBufEnd, ASN1_ERROR_OVERFLOW);
```
Let's say `len` (which is the return value from `GetLength()` and can be assumed to be malicious) is `0xFFFFFFFA` (aka `(uint32_t)-6`.  Then `bytesForLen` will be 5, `totalLen` will end up as 0, the `mWritePoint + totalLen <= mBufEnd` test will presumably succeed.  The TLV presumably does not have that much data in it, of course, but if we have a non-contiguous reader it does not know how much it has, and will keep copying data out in `GetBytes` until it runs out of data.  If it provides more data than `mBufEnd - mWritePoint`, it will overrun the end of the ASN1Writer's buffer and start writing memory we don't own, before erroring out when it discovers it does not actually have `0xFFFFFFFA` bytes of data.

 #### Proposed Solution
At a bare minimum, we need to fail out without adjusting `mWritePoint` if `GetBytes` fails.  But that won't address the EncodeHead problem or the fact that we encode a bogus length into the buffer before erroring out.

Even better might be to require a `ContiguousBufferTLVReader` and use `GetByteView`, which will at least ensure that we have all the data to back up the length we thing we are dealing with.

We could fix the EncodeHead problem by adjusting the checks to avoid overflow.  Something like this:
```
  VerifyOrReturnError(mWritePoint <= mBufEnd, ASN1_ERROR_OVERFLOW);  // Or not, if this is an invariant we maintain anyway
  VerifyOrReturnError(mBufEnd - mWritePoint >= 1 + bytesForLen, ASN1_ERROR_OVERFLOW); // Make sure we don't underflow
  VerifyOrReturnError(mBufEnd - mWritePoint - (1 + bytesForLen) >= (len != kUnknownLength ? len : 0), ASN1_ERROR_OVERFLOW);
```

**Comment by emargolis** [comment]
#9532 should fix these issues.

1. Unsafe methods `GetLength()` and `GetBytes()` were replaced with `Get(ByteSpan & v`), which performs length/data sanity checks
2. Another potential integer overflow of the `totalLen` parameter (described above) is not possible because the `len` parameter in the calculation is `int32_t`:
`totalLen = 1 + bytesForLen + (len != kUnknownLength ? len : 0);`

**Comment by bzbarsky-apple** [comment]
> is not possible because the len parameter in the calculation is `int32_t`

Huh.  Callers are passing `uint32_t` values in here.  We need to turn on `-Wconversion` for this stuff....

The main problem is fixed, I guess, with #9532.
